File: 39.组合总和.cs
/*
 * @lc app=leetcode.cn id=39 lang=csharp
 *
 * [39] 组合总和
 */

// @lc code=start
public class Solution
{
    public IList<IList<int>> CombinationSum(int[] candidates, int target)
    {
        int len = candidates.Length;
        IList<IList<int>> res = new List<IList<int>>();
        if (len == 0)
        {
            return res;
        }
        Stack<int> path = new Stack<int>();
        dfs(candidates, 0, len, target, path, res);
        return res;
    }

    public void dfs(int[] candidates, int begin, int len, int target, Stack<int> path, IList<IList<int>> res)
    {
        if (target < 0)
        {
            return;
        }
        if (target == 0)
        {
            res.Add(new List<int>(path));
            return;
        }
        for (int i = begin; i < len; i++)
        {
            path.Push(candidates[i]);
            dfs(candidates, i, len, target - candidates[i], path, res);
            path.Pop();
        }
    }
}
// @lc code=end

File: 39.组合总和.go
/*
 * @lc app=leetcode.cn id=39 lang=golang
 *
 * [39] 组合总和
 */
package main

// @lc code=start
func combinationSum(candidates []int, target int) [][]int {
	n := len(candidates)
	res := [][]int{}
	if n == 0 {
		return res
	}
	path := []int{}
	res = dfs(candidates, 0, n, target, path, res)
	return res
}

func dfs(candidates []int, begin, n, target int, path []int, res [][]int) [][]int {
	if target < 0 {
		return res
	}
	if target == 0 {
		res = append(res, append([]int(nil), path...))
	}
	for i := begin; i < n; i++ {
		path = append(path, candidates[i])
		res = dfs(candidates, i, n, target-candidates[i], path, res)
		path = path[:len(path)-1]
	}
	return res
}

// @lc code=end
